### PR TITLE
Implement API endpoint to link a package to a repo

### LINKED ISSUE #21062: Ability to link package to a repo from api
### Feature Description

We can push package from the api but we can't link it to a repo afterward.
I think it's could be useful to manage this from the api

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by sclu1034**
Closes #21062.

### CODE REVIEW

File: services/packages/packages.go [hunk 1]
```diff
@@ -554,3 +556,35 @@ func RemoveAllPackages(ctx context.Context, userID int64) (int, error) {
 	}
 	return count, nil
 }
+
+func LinkPackage(ctx context.Context, doer *user_model.User, p *packages_model.Package, repoID int64) error {
```
Comment (KN4CK3R) [inline services/packages/packages.go]:
```suggestion
func LinkPackageToRepository(ctx context.Context, doer *user_model.User, p *packages_model.Package, repoID int64) error {
```

File: routers/api/v1/packages/package.go [hunk 2]
```diff
@@ -211,3 +211,52 @@ func ListPackageFiles(ctx *context.APIContext) {
...
+	repoID := ctx.FormInt64("repo")
+	packageType := ctx.Params("type")
+	name := ctx.Params("name")
+	pkg, err := packages.GetPackageByName(ctx, ctx.ContextUser.ID, packages.Type(packageType), name)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "LinkPackage", err)
+		return
+	}
+
+	err = packages_service.LinkPackage(ctx, ctx.Doer, pkg, repoID)
```
Comment (KN4CK3R) [inline routers/api/v1/packages/package.go]:
```suggestion
	err = packages_service.LinkPackageToRepository(ctx, ctx.Doer, pkg, repoID)
```

File: routers/api/v1/packages/package.go [hunk 3]
```diff
@@ -211,3 +211,52 @@ func ListPackageFiles(ctx *context.APIContext) {
...
+		ctx.Error(http.StatusInternalServerError, "LinkPackage", err)
+		return
+	}
+
+	err = packages_service.LinkPackageToRepository(ctx, ctx.Doer, pkg, repoID)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "LinkPackage", err)
+		return
+	}
+	ctx.Status(http.StatusNoContent)
```
Comment (KN4CK3R) [inline routers/api/v1/packages/package.go]:
```suggestion
	pkg, err := packages.GetPackageByName(ctx, ctx.ContextUser.ID, packages.Type(ctx.Params("type")), ctx.Params("name"))
	if err != nil {
		if errors.Is(err, utils.ErrNotExist) {
			ctx.Error(http.StatusNotFound, "GetPackageByName", err)
		} else {			
			ctx.Error(http.StatusInternalServerError, "GetPackageByName", err)
		}
		return
	}

	err = packages_service.LinkPackageToRepository(ctx, ctx.Doer, pkg, ctx.FormInt64("repo"))
	if err != nil {
		ctx.Error(http.StatusInternalServerError, "LinkPackage", err)
		return
	}
	ctx.Status(http.StatusNoContent)
```
And it needs an additional import for `errors` then.

**Comment by denyskon** [comment]
@sclu1034 Cloud you update your branch and resolve merge conflicts?

**Comment by sclu1034** [comment]
Rebased on current `main`

File: services/packages/packages.go [hunk 4]
```diff
@@ -657,3 +659,35 @@ func RemoveAllPackages(ctx context.Context, userID int64) (int, error) {
...
+			if err != nil {
+				return fmt.Errorf("Error getting repository permissions for %d on %d: %w", doer.ID, repo.ID, err)
+			}
+
+			canWrite = perms.CanWrite(unit.TypePackages)
+		}
+
+		if !canWrite {
+			return fmt.Errorf("No permissions to link this package and repository")
+		}
```
Comment (denyskon) [inline services/packages/packages.go]:
```suggestion
	  perms, err := access_model.GetUserRepoPermission(ctx, repo, doer)
	  if err != nil {
		  return fmt.Errorf("Error getting repository permissions for %d on %d: %w", doer.ID, repo.ID, err)
	  }

		if !perms.CanWrite(unit.TypePackages) {
			return fmt.Errorf("No permissions to link this package and repository")
		}
```

It is also possible that packages are disabled for this repo, so I think we should check for all users, not just if user isn't the owner.

		
```

File: services/packages/packages.go [hunk 5]
```diff
@@ -657,3 +659,35 @@ func RemoveAllPackages(ctx context.Context, userID int64) (int, error) {
...
+
+		repoID = repo.ID
```
Comment (denyskon) [inline services/packages/packages.go]: Why is this needed?

**Comment by VAllens** [comment]
Goooooooooooooooooooooooooooooooooooooooooood !!!

File: services/packages/packages.go (hunk 5 above)
Comment (sclu1034) [reply services/packages/packages.go]: Probably a leftover from when I was working on this

File: services/packages/packages.go [hunk 6]
```diff
@@ -657,3 +659,29 @@ func RemoveAllPackages(ctx context.Context, userID int64) (int, error) {
 	}
 	return count, nil
 }
+
+func LinkPackageToRepository(ctx context.Context, doer *user_model.User, p *packages_model.Package, repoID int64) error {
+	if repoID != 0 {
+		repo, err := repo_model.GetRepositoryByID(ctx, repoID)
```
Comment (KN4CK3R) [inline services/packages/packages.go:665]: This MUST check if the repo belongs to the package owner. And there should be a test for linking a repo of another user.

File: routers/api/v1/api.go [hunk 7]
```diff
@@ -1412,6 +1412,7 @@ func Routes() *web.Route {
 				m.Delete("", reqToken(), reqPackageAccess(perm.AccessModeWrite), packages.DeletePackage)
 				m.Get("/files", reqToken(), packages.ListPackageFiles)
 			})
+			m.Post("/{type}/{name}/link", reqToken(), reqPackageAccess(perm.AccessModeWrite), packages.LinkPackage)
```
Comment (KN4CK3R) [inline routers/api/v1/api.go]: Please replace this block with this:
```go
			m.Group("/{type}/{name}", func() {
				m.Group("/{version}", func() {
					m.Get("", packages.GetPackage)
					m.Delete("", reqPackageAccess(perm.AccessModeWrite), packages.DeletePackage)
					m.Get("/files", packages.ListPackageFiles)
				})
				m.Post("/link", reqPackageAccess(perm.AccessModeWrite), packages.LinkPackage)
			}, reqToken())
```

File: routers/api/v1/packages/package.go [hunk 8]
```diff
@@ -213,3 +214,54 @@ func ListPackageFiles(ctx *context.APIContext) {
...
+	// - name: name
+	//   in: path
+	//   description: name of the package
+	//   type: string
+	//   required: true
+	// - name: repo
+	//   in: query
+	//   description: ID of the repository to link
+	//   type: integer
+	//   required: true
```
Comment (KN4CK3R) [inline routers/api/v1/packages/package.go]: Would be nice if you can't just pass the repository id (which you don't have usually) but the repo name.
Could have two parameters `repo_id=...` and `repo_name=...`. The endpoint should choose the matching repo then. Could prefer the id over the name.

**Review by KN4CK3R — CHANGES_REQUESTED** [review]

File: services/packages/packages.go (hunk 6 above)
Comment (sclu1034) [reply services/packages/packages.go:665]: To avoid further back and forth on permissions, and me doing best-effort guesses, can you give me a detailed run down on who should be allowed to do what?

File: routers/api/v1/packages/package.go (hunk 8 above)
Comment (sclu1034) [reply routers/api/v1/packages/package.go]: Not particularly proud of how my code parsing those two parameters turned out, but I didn't find any useful utilities.
Any ideas to improve [`221dcdc`](https://github.com/go-gitea/gitea/pull/23851/commits/221dcdc4beb7bf79bcd1ad060172800aa28d28a9)?

**Comment by KoltesDigital** [comment]
@KN4CK3R can you please have a look at the author's questions?

File: routers/api/v1/packages/package.go [hunk 9]
```diff
@@ -213,3 +218,96 @@ func ListPackageFiles(ctx *context.APIContext) {
...
+			ctx.Error(http.StatusNotFound, "GetPackageByName", err)
+		} else {
+			ctx.Error(http.StatusInternalServerError, "GetPackageByName", err)
+		}
+		return
+	}
+
+	var repoID int64
+
+	formRepoID := ctx.FormString("repo_id")
```
Comment (lunny) [inline routers/api/v1/packages/package.go:271]: Do we really want to expose the repository's id? I don't think we should do that. Using `OwnerName` and `RepoName` should be better.

Comment (sclu1034) [reply routers/api/v1/packages/package.go:271]: I'm not sure what you mean by "exposing" in this context. The existing `GET /repos/{owner}/{repo}` endpoint makes the `id` value available in its response, so that information is already "exposed" to every consumer of the public API.

What's the harm in giving the user a choice which identifier they can run this new endpoint against?

Comment (KoltesDigital) [reply routers/api/v1/packages/package.go:271]: @sclu1034 I think last comments refer to the fact that most, if not all the API, refer to orgs and repos using their name, and not their id. At least I already played with the API and I don't recall I ever sent requests using the repo ids. It just feels more natural to use the repo name instead.

Moreover, the web frontend only allow linking to a repository of the same owner. If that is by design, you would have to drop the owner part in the repo name.
EDIT: from @KN4CK3R's previous comment, it is indeed by design.

Just my last two cents on the endpoint: there could only be zero or one link to a repository, so I'd rather use PUT and DELETE.

Comment (lunny) [reply routers/api/v1/packages/package.go:271]: > Moreover, the web frontend only allow linking to a repository of the same owner. If that is by design, you would have to drop the owner part in the repo name.

I think this is by design. It's weird to link a package to another owner's repository.

Comment (KoltesDigital) [reply routers/api/v1/packages/package.go:271]: @lunny well I'm still wondering. I develop some Unity (game engine) packages. Distribution relies on NPM protocol, although packages are obviously not meant for Nodejs (it's C#). And I also develop some Nodejs packages. But Gitea only provides a single NPM registry per owner. So it seems to me the only way to cleanly separate both usages is to make a dedicated "owner", so to have two registries. But this is only for distribution. Development still occurs in the same org.

**Comment by sclu1034** [comment]
Closing, as evidently, with months of silence and accumulating conflicts while waiting for comments on my questions, this is going nowhere.
